Reboot Class Diagram and rename client libraries (#176)
eclipse-uprotocol/up-spec: languages.adoc

SPDX-License-Identifier: Apache-2.0
----

== Overview 

> A uProtocol Language specific libraries (_libraries_), are written for a given programming language (ex. Java, C++, etc...) to implement the protocol and are the primary library that is used by applications and services (uEs) to communicate with other uEs. The _libraries_ declare the API for the uProtocol Transport Layer (uP-L1) and implement the uProtocol Communication Layer API (uP-L2).

Reviewer: ```suggestion
uProtocol Language specific libraries (_libraries_) are written for a given programming language (ex. Java, C++, etc...) to implement the protocol and are the primary library that is used by applications and services (uEs) to communicate with other uEs. The _libraries_ declare the API for the uProtocol Transport Layer (uP-L1) and implement the uProtocol Communication Layer API (uP-L2).
```

== Requirements

=== Content

> [oft-sid="req~up-language-api~1"]

Reviewer: Based on my current implementation of the AsciiDoc support for OpenFastTrace, this should also include the `.specitem` label:
```
[.specitem,oft-sid="req~up-language-api~1"]
```
This is true for the other requirements as well.

Author: I fixed them all now

--
* *MUST* implement serializers, validators, and builders for the link:../basics/README.adoc[uProtocol data types] using the link:../up-core-api/README.adoc[up-core-api data model].
--

[oft-sid="req~up-language-transport-api~1"]
--
* *MUST* declare the language specific link:up-l1/README.adoc[uProtocol Transport Layer interface], such that the interface can be implemented by the transport libraries.
--

[oft-sid="req~up-language-comm-api~1"]
--

> * *MUST* provide an implementation of the link:up-l2/README.adoc[uProtocol Communication Layer (uP-L2) `API].

Reviewer: ```suggestion
* *MUST* declare the language specific link:up-l2/README.adoc[uProtocol Communication Layer (uP-L2) API] and provide a default implementation based on the (abstract) Transport Layer API.
```

--

[oft-sid="req~up-language-naming~1"]
--
* *MUST* be named according to pattern `up-[LANGUAGE]`
--

[oft-sid="req~up-language-structure~1"]
--
* The top level namespace (package name) *MUST* contain `uprotocol` and *SHOULD* contain `org.eclipse.uprotocol` 
--

[oft-sid="req~up-language-documentation~1"]
--

>  * *MUST* contain a `README.adoc` that explains how to build and use the library. 

Reviewer: what about API docs?

Author: API documentation is in the code and not in the README usually.

--

=== Test & CI

[oft-sid="req~up-language-test~1"]
--
 * *MUST* be able to run the test suite such that merge requests (commits) cannot be merged till tests are passed
--

[oft-sid="req~up-language-ci-build~1"]
--
 * *MUST* have CI workflow that verifies pull requests compiles before they can be merged
--

[oft-sid="req~up-language-ci-test~1"]
--

>  * *SHOULD* cover 100% of all code and specifications defined in <<Source>>

Reviewer: is this a requirement for the CI system or for the (unit) test suite itself?

Author: unit tests and then CI runs to ensure it is 100% covered.

Reviewer: If I am not mistaken, the link target does not exist.

Author: Ok I fixed it, thanks for catching that

--

[oft-sid="req~up-language-ci~3"]
--

>  * *SHOULD* have lint & documentation coverage checks

Reviewer: what is *documentation coverage*?

Author: ```suggestion
 * *SHOULD* have linter for static code analysis as well as a tool that verifies APIs are properly documented
```

> [.specitem,oft-sid="req~up-language-ci~3"]

Reviewer: do you want this to be revision `3`?

--

=== Build

[oft-sid="req~up-language-build-sys~1"]
--
 * *SHOULD* employ an existing build system that is commonly used for the given programming language
--

[oft-sid="req~up-language-build-deps~1"]
--

>  * *MUST NOT* duplicate external dependencies but pull them into the build process

Reviewer: ```suggestion
 * *MUST* resolve and include external dependencies as part of the build process only
```

> [.specitem,oft-sid="req~up-language-build-sys~1",oft-needs="impl,utest"]

Reviewer: Any idea how a unit test for this would/could look like? Maybe requiring an `impl` would be enough here? But we can change that later ...
